types/storage_key.go

```diff
@@ -40,32 +40,52 @@ func NewStorageKey(b []byte) StorageKey {
 
 // CreateStorageKey uses the given metadata and to derive the right hashing of module, fn names and keys to create a
 // hashed StorageKey
-func CreateStorageKey(meta Metadata, module string, fn string, key []byte) (StorageKey, error) {
-	var fnMeta *StorageFunctionMetadata
-	for _, m := range meta.Metadata.Modules {
-		if m.Prefix == module {
-			for _, s := range m.Storage {
-				if s.Name == fn {
-					fnMeta = &s //nolint:scopelint
-					break
+func CreateStorageKey(meta *Metadata, module string, fn string, key []byte) (StorageKey, error) {
+	var hasher hash.Hash
+	var err error
+
+	var found bool
+	if meta.IsMetadataV4 {
```

Comment:
can we refactor these into functions for version handling? this code got even uglier from my code before :|

Reply:
You are absolutely right – I'll do that in my latest PR. I think the best way would actually be to expose a Metadata interface that the different versions need to implement, i. e. 

```
type Metadata interface {
	FindCallIndex(call string) (CallIndex, error)
	FindEventNamesForEventID(eventID EventID) (Text, Text, error)
	FindStorageKeyHasher(module string, fn string) (hash.Hash, error)
}
```

That would prevent unsuported metadata errors at compile time and simplify the code.

But it's a major refactoring, so I'd suggest we just go ahead with what we have right now – what do you think?

Reply:
Refactored here: https://github.com/centrifuge/go-substrate-rpc-client/pull/28/commits/8b2f69a9f8930d3f5932d666e31eb1c8aeac1ab3